Refactor PropertyTester structure

ValerianRey opened the pull request:
* Transform property testers in _property_testers.py into assertion functions in _asserts.py
* Add n_runs parameter to assert_linear_under_scaling
* Update aggregator tests to parameterize and use the new assertion functions rather than property testers
* Fix dtype issues
* Add test_permutation_invariant to ConFIG and IMTLG
* Update tolerances

* Refactor matrix sampling functions into classes and move them to _matrix_samplers.py
* Add dtype parameter to the constructor of MatrixSamplers
* Add rng parameter to the call method of MatrixSamplers

Note that we decided to have default values for some parameters of the assertion functions (atol, rtol, n_runs and threshold), so that we still have the benefit of having most aggregators tested with the same tolerances, for ease of maintenance (as we don't want to update each aggregator's tolerance individually each time inputs, hardware or cuda drivers change). We have decided that only UPGrad and DualProj would use custom (and tight) tolerances, so that we're aware if we make a regression for them.

For some reason, the permutation_invariance tests now pass on cuda for ConFIG and IMTLG, so they now test it too. This could be due to a change in the cuda implementation of pinv.

ValerianRey commented:
@PierreQuinton this is not finished yet, but do you agree with the structure so far?

The idea is to:
- Get rid of test classes since they're not really necessary and it makes it harder to run exactly the tests that you want
- Have the functions defined in _property_testers be as parametrizable as possible: for instance, they take atol, rtol, n_perms and threshold parameters (depending on what is tested). This makes it possible to have a atol and rtol per aggregator for instance.
- They also act on a single matrix, so it's now the responsibility of the caller to define on which matrices these tests will be run. Previously, there was a default set of matrices on which those tests would be run, and some aggregators (that didn't work with those matrix shapes) had to override this parametrization, which was not clean IMO.

ValerianRey commented:
I think the next steps (probably in other pull requests) are:
- Use rtol a lot more often and reduce tolerance as much as possible
- Improve matrix generation: I would like to have MatrixSampler classes in _matrix_samplers.py, have them take a dtype and an rng state, and have the actual inputs still in _inputs.py (so that both remain separated).

PierreQuinton commented on tests/unit/aggregation/_property_testers.py:
```diff
@@ -1,144 +1,95 @@
-        for _ in range(PermutationInvarianceProperty.N_PERMUTATIONS):
-            permuted_matrix = PermutationInvarianceProperty._permute_randomly(matrix)
-            permuted_vector = aggregator(permuted_matrix)
+    for _ in range(n_perms):
```
So either we add a for loop there or we remove it here

ValerianRey replied:
Because of this fixture in conftest.py:
```python
@fixture(autouse=True)
def fix_randomness() -> None:
    rand.seed(0)
    torch.manual_seed(0)
    torch.use_deterministic_algorithms(True)
```
The randomness is reset between each test, and even between each parametrization of each test.

For instance, for all matrices of 9 rows, we will get the following 5 random permutations:
```python
tensor([8, 0, 2, 3, 7, 1, 4, 5, 6])
tensor([1, 4, 0, 5, 3, 8, 7, 6, 2])
tensor([0, 5, 6, 2, 1, 8, 4, 7, 3])
tensor([8, 2, 3, 0, 1, 5, 7, 6, 4])
tensor([8, 1, 5, 6, 3, 0, 2, 7, 4])
```

Without a loop, we would always just test the first permutation, on all matrices of 9 rows:
```python
tensor([8, 0, 2, 3, 7, 1, 4, 5, 6])
```
I think this could be insufficient.

I would thus prefer to keep having a for-loop here.

The same can also be said about the linear under scaling test. Currently, the values obtained for c1, c2, alpha and beta are always the same when the matrix has the same number of rows.
For instance, we have:
```python
c1: tensor([0.4963])
c2: tensor([0.7682])
alpha: tensor(0.0885)
beta: tensor(0.1320)
```
for matrices of 1 row, and
```python
c1: tensor([0.4963, 0.7682, 0.0885, 0.1320])
c2: tensor([0.3074, 0.6341, 0.4901, 0.8964])
alpha: tensor(0.4556)
beta: tensor(0.6323)
```
for matrices of 4 rows. Notice how the value of c1 for matrices of 4 rows contains the values of c1, c2, alpha and beta obtained for matrices of 1 row.

So my guess is that we should improve this test by also adding a for-loop there. This makes the whole testing about 1 sec longer on my machine.

PierreQuinton commented on tests/unit/aggregation/_property_testers.py:
```diff
@@ -1,144 +1,95 @@
-        assert norm > 1e-03
+    vector = aggregator(matrix)
+    norm = vector.norm().item()
+    assert norm > threshold
```
threshold is a rtol (or maybe atol), maybe we should use assert_close(vector.norm(), zero)

ValerianRey replied:
Actually we want to test that norm is not close to zero, rather than the opposite.

We would need to do:
```python
with raises(AssertionError):
    assert_close(norm, zero, rtol=0., atol=threshold)
```
or 
```python
assert not torch.allclose(norm, zero, rtol=0., atol=threshold)
```

I think both solutions just reduce readability, so I'll keep this test as it is.

PierreQuinton reviewed (commented):
So far so good. Shiuld we also fine tune the tolerences in this PR?

ValerianRey commented:
> So far so good. Shiuld we also fine tune the tolerences in this PR?

Tolerances will depend on the inputs. We should first rework the way inputs are generated before fine-tuning tolerances. We can do everything in the same PR though.

ValerianRey commented:
@PierreQuinton What do you think of the new MatrixSampler structure? The idea here is to make very powerful and generic classes that could be reused very easily in arena for example (for now with copy-paste of the whole file, but maybe one day through a shared dependency).

I added an rng parameter, which allows using a generator for these generations, and which is the standard in torch random samplers. I named it rng instead of generator to make it shorter. I made this a parameter of the call method of the samplers, and I gave a default value of None, to match what torch does (so that we could use these classes without the generator if needed).

Lastly, I made the dtype default to torch.float32 in MatrixSampler. Again, the goal is to make it easier to generate matrices and to try to stick with torch's defaults.

Note that I also tried running the tests with torch.float64 matrices (something that we never did as far as I remember), on both cpu and cuda. A few tests failed due to forgetting specifying the dtype in some locations, but I fixed that in e2c054bc8e4950a113b5d334d89bcf0eeb90ec92. With this change, all tests pass on float64 on cpu, and all except 1 pass on cuda (by a very short margin, so it's not a real issue). It's good to know that we now have a very easy way to test on float64 (although I wouldn't do it automatically as it would cost time).

For me the last remaining step is reworking the tolerance params. I think it will be a lot of work to ensure that the rtol and atol are always as tight as possible. It requires a few hours of upfront work, and then it requires some extra work each time an aggregator changes, the inputs change, or the hardware changes (or even a hardware driver). So in my opinion we should have default values for rtol and atol (those we had before for instance, or with more emphasis on the rtol), and make them tighter (or less tight) for the most important aggregators (I would say only UPGrad, maybe DualProj too) when needed.

PierreQuinton commented:
Everything is very nice. I agree with having a specific atol/rtol for some specific aggregators only, so I think we should set precision for those, then tighten the precision for all others (for instance LUS should get more precise if we remove `UPGrad`).

ValerianRey commented:
For me this is ready to merge.

PierreQuinton reviewed (approved):
LGTM, it might be worth having different values for rtol and atol (make it almost tight) in the future, we should open an issue.

ValerianRey commented:
> LGTM, it might be worth having different values for rtol and atol (make it almost tight) in the future, we should open an issue.

I don't think tolerances should be too tight. They should be tight enough so that we realize it whenever we make a significant regression, but they should also be large enough so that most changes to hardware do not break the tests. This is because we have no guarantee about what machines GitHub will provide us with for actions, and we also want to be robust to ourselves' change of machine or future contributors own hardware.

Also, there's an infinite number of possibilities to be tight if we select both rtol and atol. It's not like assert_close checks that we're tight both relatively and absolutely. So we could make things tighter, less tight, more focused on relative tightness or more focused on absolute tightness, but we can't make things extremely tight both relatively and absolutely. So for me there's no need to open an issue before we have an actual problem (for instance a test stops passing for some reason).

PierreQuinton commented:
You are right, however we could say look at the worst atol difference for a given test and make the value twice that. We could do the same for rtol independently.

ValerianRey commented:
> You are right, however we could say look at the worst atol difference for a given test and make the value twice that. We could do the same for rtol independently.

I like the idea of chosing double the tightest value, but let's only spend time on that when we really need to (when some tests start failing for some reason).
